File: 스택/9012/9012.cpp
#include <iostream>
#include <stack>
#include <string>
using namespace std;

int main()
{
   stack<bool> s;
   int n;
   cin>>n;

   for (int i = 0; i < n; ++i)
   {
      string t;
      cin>>t;
      for (int j = 0; j < t.size(); ++j)
      {
         if (s.empty() or t[j] == '(')
            s.push(t[j] - '(');
         else if (!s.top())
            s.pop();
         else
            s.push(true);
      }
      if (s.empty())
         cout<<"YES\n";
      else
         cout<<"NO\n";
      while (!s.empty())
         s.pop();
   }

   return 0;
}

File: 스택/9012/9012.py
n = int(input())
arr = []

for i in range(n):
   s = input()
   for j in s:
      if len(arr) == 0 or j == '(':
         arr.append(True if j == '(' else False)
      elif arr[-1]:
         arr.pop()
      else:
         arr.append(False)
   if len(arr) == 0:
      print("YES")
   else:
      print("NO")
   arr.clear()